File: source/python/http_server.py
import threading
import queue

request_q = queue.Queue(maxsize=10)
response_q = queue.Queue(maxsize=10)

def server():
    for _ in range(3):
        req = request_q.get()
        method = req["method"]
        path = req["path"]
        status = 200
        if path == "/":
            body = "Hello, World!"
        elif path == "/about":
            body = "About Page"
        else:
            status = 404
            body = "Not Found"
        response_q.put({"status": status, "body": body, "method": method, "path": path})

def client():
    requests = [
        {"method": "GET", "path": "/"},
        {"method": "GET", "path": "/about"},
        {"method": "GET", "path": "/missing"},
    ]
    for req in requests:
        request_q.put(req)
        res = response_q.get()
        print(f"[simulated] {res['method']} {res['path']} -> {res['status']} {res['body']}")

srv = threading.Thread(target=server)
cli = threading.Thread(target=client)

srv.start()
cli.start()
cli.join()
srv.join()

File: source/rust/http_server.rs
use std::sync::mpsc;
use std::thread;

struct Request { method: String, path: String }
struct Response { status: u32, body: String, method: String, path: String }

fn handle(req: Request) -> Response {
    let (status, body) = match req.path.as_str() {
        "/" => (200, "Hello, World!".to_string()),
        "/about" => (200, "About Page".to_string()),
        _ => (404, "Not Found".to_string()),
    };
    Response { status, body, method: req.method, path: req.path }
}

fn main() {
    let (req_tx, req_rx) = mpsc::channel::<Request>();
    let (res_tx, res_rx) = mpsc::channel::<Response>();

    // Server
    let server = thread::spawn(move || {
        for req in req_rx.into_iter().take(3) {
            res_tx.send(handle(req)).unwrap();
        }
    });

    // Client
    let client = thread::spawn(move || {
        let requests = vec![
            Request { method: "GET".to_string(), path: "/".to_string() },
            Request { method: "GET".to_string(), path: "/about".to_string() },
            Request { method: "GET".to_string(), path: "/missing".to_string() },
        ];
        for req in requests {
            req_tx.send(req).unwrap();
            let res = res_rx.recv().unwrap();
            println!("[simulated] {} {} -> {} {}", res.method, res.path, res.status, res.body);
        }
    });

    client.join().unwrap();
    server.join().unwrap();
}
